File: Sol30_Substring_With_Concentration_of_All_Words/Solution30.java
import java.util.*;

class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        int n = s.length();
        int wordLen = words[0].length();
        int sentenceDesireLen = wordLen * words.length;

        List<Integer> list = new ArrayList<>();
        HashMap<String, int[]> map = new HashMap<>();

        for (String word : words) {
            map.putIfAbsent(word, new int[] {0, 0});
            map.get(word)[0]++;
        }

        for (int i = 0; i < wordLen; i++) {
            int left = i, right = i;

            while (right <= n - wordLen &&
                   left <= n - sentenceDesireLen) {

                String word = s.substring(right, right + wordLen);

                if (map.containsKey(word)) {
                    int[] freq = map.get(word);

                    while (freq[0] == freq[1]) {
                        String prevWord =
                            s.substring(left, left + wordLen);

                        map.get(prevWord)[1]--;
                        left += wordLen;
                    }

                    freq[1]++;
                } else {
                    while (left != right) {
                        String prevWord =
                            s.substring(left, left + wordLen);

                        map.get(prevWord)[1]--;
                        left += wordLen;
                    }

                    left += wordLen;
                }

                right += wordLen;

                if (right - left == sentenceDesireLen) {
                    list.add(left);
                }
            }

            while (left != right) {
                String prevWord =
                    s.substring(left, left + wordLen);

                map.get(prevWord)[1]--;
                left += wordLen;
            }
        }

        Collections.sort(list);
        return list;
    }


    public static void main(String[] args) {

        Solution solution = new Solution();

        System.out.println("Example 1:");
        String s1 = "barfoothefoobarman";
        String[] words1 = {"foo", "bar"};

        System.out.println("Input: " + s1 +
                           " " + Arrays.toString(words1));
        System.out.println("Output: " +
                           solution.findSubstring(s1, words1));


        System.out.println("\nExample 2:");
        String s2 = "wordgoodgoodgoodbestword";
        String[] words2 = {"word", "good", "best", "word"};

        System.out.println("Input: " + s2 +
                           " " + Arrays.toString(words2));
        System.out.println("Output: " +
                           solution.findSubstring(s2, words2));


        System.out.println("\nExample 3:");
        String s3 = "barfoofoobarthefoobarman";
        String[] words3 = {"bar", "foo", "the"};

        System.out.println("Input: " + s3 +
                           " " + Arrays.toString(words3));
        System.out.println("Output: " +
                           solution.findSubstring(s3, words3));
    }
}

File: Sol30_Substring_With_Concentration_of_All_Words/Solution30.py
from collections import defaultdict
from typing import List

class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        if not s or not words:
            return []

        word_count = defaultdict(int)
        for word in words:
            word_count[word] += 1

        substr_len = len(words) * len(words[0])
        word_len = len(words[0])
        result = []

        for i in range(len(s) - substr_len + 1):
            seen = defaultdict(int)

            for j in range(i, i + substr_len, word_len):
                word = s[j:j + word_len]

                if word in word_count:
                    seen[word] += 1

                    if seen[word] > word_count[word]:
                        break
                else:
                    break
            else:
                result.append(i)

        return result


# Examples

solution = Solution()

print("Example 1:")
s = "barfoothefoobarman"
words = ["foo", "bar"]
print("Input:", s, words)
print("Output:", solution.findSubstring(s, words))

print("\nExample 2:")
s = "wordgoodgoodgoodbestword"
words = ["word", "good", "best", "word"]
print("Input:", s, words)
print("Output:", solution.findSubstring(s, words))

print("\nExample 3:")
s = "barfoofoobarthefoobarman"
words = ["bar", "foo", "the"]
print("Input:", s, words)
print("Output:", solution.findSubstring(s, words))